### PR TITLE
Add get recommended outbound channels endpoint

### PR DESCRIPTION
**Pull request by 0xBEEFCAF3**

### CODE REVIEW

**Comment by nickfarrow** [comment]
Very very nice, just leaving some non-specific comments for my thoughts on direction

File: src/lib.rs [hunk 1]
```diff
@@ -6,4 +6,5 @@ pub mod scheduler;
 pub mod http;
 pub mod args;
 pub mod lnd;
+pub mod mempool;
```
Comment (nickfarrow) [inline src/lib.rs:9]: maybe replace `mempool` with `recommend` or something less specific (we will likely want other feeds for these recommendations, im working on one now and will PR to your PR.)

File: src/mempool.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+pub async fn get_recommended_channels() -> Result<Vec<String>, hyper::http::Error> {
+    let https = hyper_tls::HttpsConnector::new();
+    let client = Client::builder()
+        .build::<_, hyper::Body>(https);
+    let uri: Uri = "https://mempool.space/api/v1/lightning/nodes/rankings/connectivity".parse().unwrap();
+    let res = client.get(uri).await.unwrap();
+
+    // Recommend 6 high liquidity channels
+    // Recommend 4 smaller edge nodes
```
Comment (nickfarrow) [inline src/mempool.rs:36]: I am thinking that this too many to start with (we ourselves have only tested 2 channels afaik).

How do you feel about suggesting 3:
1x Routing node
1x edge node
1x Nolooking special suggestion node (e.g. HRF)

Comment (nickfarrow) [reply src/mempool.rs:36]: (im writing a commit to fetch nolooking special suggested node)

File: src/mempool.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    let body_str = body_to_string(res.into_body()).await;
+    let high_channels: ConnectivityResponse = serde_json::from_str(&body_str).unwrap();
+    let mut high_cap_nodes = high_channels.clone(); 
+    high_cap_nodes.sort_by(|a, b| a.capacity.partial_cmp(&b.capacity).unwrap());
+    
+    let base_uri = "https://mempool.space/api/v1/lightning/nodes";
+    // Create a vector of lightning addresses
+    // Prefer onion address is accessible
```
Comment (nickfarrow) [inline src/mempool.rs:44]: I'm split on this one, I am thinking we could lean towards IP for the routing node and use onions for the edge nodes

Comment (0xBEEFCAF3) [reply src/mempool.rs:44]: Given the tor DDOS attacks I think we should always lead with IP for now

**Comment by 0xBEEFCAF3** [comment]
This how we lookin

Screen Shot 2022-11-10 at 8 53 24 AM

**Comment by DanGould** [comment]
What if we default to those nodes so there's no need to hit "use?" Then if they're connected, they don't show up again

> The answer is not more options. If you feel compelled to add a preference that's exposed to the user, it's very possible you've made a wrong turn somewhere.
https://github.com/signalapp/Signal-iOS/blob/main/CONTRIBUTING.md#development-ideology

**Comment by 0xBEEFCAF3** [comment]
> What if we default to those nodes so there's no need to hit "use?" Then if they're connected, they don't show up again

Defaulting to recommendations could make sense. But we still want the recommendations to be dynamic

File: public/index.html [hunk 4]
```diff
@@ -82,8 +111,10 @@ <h2>PayJoin here to open these channels</h2>
 
 		</main>
 		<script type="text/javascript">
-			function add_channel() {
-				let channels = document.getElementById("channels");
+			let recommendations;
+			var channels = 1;
```
Comment (nickfarrow) [inline public/index.html:115]: think we can delete this `channels = 1`

File: public/index.html [hunk 5]
```diff
@@ -82,8 +111,10 @@ <h2>PayJoin here to open these channels</h2>
...
+			function add_channel(address=""){
```
Comment (nickfarrow) [inline public/index.html:116]: Doesn't add the new channel ID for me. Setting the `value` if `address != ""`:
```
function add_channel(address=""){
	var channels = document.getElementById("channels");
	let div = document.createElement('div');
	if (address == "") {
		div.innerHTML = `<input type="text" name="channels[${channels.length + 1}][node]" aria-labelledby="destinationLabel" placeholder="node@host:port" span="1" required><input type="number" name="channels[${channels.length + 1}][amount]" min="20000" step="1" aria-labelledby="capacityLabel" span="1" required>`;
	} else {
		div.innerHTML = `<input type="text" value="${address}" name="channels[${channels.length + 1}][node]" aria-labelledby="destinationLabel" placeholder="node@host:port" span="1" required><input type="number" name="channels[${channels.length + 1}][amount]" min="20000" step="1" aria-labelledby="capacityLabel" span="1" required>`;
	}
	channels.appendChild(div.firstChild);
	channels.appendChild(div.lastChild);
}
```

File: public/index.html [hunk 6]
```diff
@@ -101,6 +132,37 @@ <h2>PayJoin here to open these channels</h2>
 				channels.removeChild(channels.children[channels.children.length - 2]);
 			}
 
+			async function get_recommended_channels(){
+				document.getElementsByClassName('recommendation-button-wrapper')[0].classList.add('invisible')
```
Comment (nickfarrow) [inline public/index.html:136]: Could make this change button text to `Refresh Recommended Channels` and give random ones from the api. Or maybe leave for another PR

**Comment by nickfarrow** [comment]
Awesome!! I think we should leave it up to the users as which ones they add. If we make recommend channels spit out semi random ones that would be super cool

File: src/recommend.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    #[serde(rename = "first_seen")]
+    pub first_seen: i64,
+    #[serde(rename = "updated_at")]
+    pub updated_at: i64,
+    pub color: String,
+    pub sockets: String,
+    #[serde(rename = "as_number")]
+    pub as_number: i64,
+    #[serde(rename = "city_id")]
+    pub city_id: Value,
```
Comment (nickfarrow) [inline src/recommend.rs:33]: Sometimes these fields are `null`. Can we skip deserializing the ones we don't need? Or use Option

```
{"public_key":"028624b5cb872511d8b23a25ec351f6201af7e6b640edc0888fce05a8dede2bc8a","alias":"[SatSale]|LabNode","first_seen":1653621833,"updated_at":1666958783,"color":"#50beef","sockets":"24nrvuglk652fenrjorggbi4d53rly56w7xhetux3fzf6wbbxdahi4yd.onion:8888","as_number":null,"city_id":null,"country_id":null,"subdivision_id":null,"longitude":null,"latitude":null,"iso_code":null,"as_organization":null,"city":null,"country":null,"subdivision":null,"active_channel_count":1,"capacity":"1000000","opened_channel_count":1,"closed_channel_count":2}
```

Comment (nickfarrow) [reply src/recommend.rs:33]: e.g. like in https://github.com/chaincase-app/nolooking/pull/61/commits/0c0e97d597e74ee45c10828459a671e75af70b16

File: public/style.css [hunk 8]
```diff
@@ -93,3 +93,38 @@ input {
...
+    text-shadow: 10px 10px 15px rgb(0, 0, 0);
+    padding: 0px 0px 0px 10px;
+    color: whitesmoke;
+    background-clip: text;
+    -webkit-background-clip: text; 
+}
+
+.recommendation-node-button-wrapper {
+    padding: 0px 0px 0px 10px; 
+}
```
Comment (DanGould) [inline public/style.css:130]: Why is everything custom? Shouldn't we be using the framework theme we're using and modifying it if we want changes? 

https://selekkt.dk/skelet/v3/

File: src/http.rs [hunk 9]
```diff
@@ -125,6 +127,13 @@ async fn handle_schedule(
     Ok(response)
 }
 
+async fn handle_get_recommendations() -> Result<Response<Body>, HttpError> {
+    let nodes = get_recommended_channels().await.unwrap();
+    let mut response = Response::new(Body::from(serde_json::to_string(&nodes).unwrap()));
+    response.headers_mut().insert(hyper::header::CONTENT_TYPE, "application/json".parse().unwrap());
```
Comment (DanGould) [inline src/http.rs:133]: TODO handle HttpErrors

File: src/recommend.rs (hunk 7 above)
Comment (DanGould) [reply src/recommend.rs:33]: Yes, unused fields should be excluded from our struct. serde will ignore them. No need for Options fields except to represent truly optional values.

Also. What's going on with the renames?

File: public/index.html [hunk 10]
```diff
@@ -101,6 +132,37 @@ <h2>PayJoin here to open these channels</h2>
...
+				} catch (error) {
+					console.error(error)
+					alert("Something went wrong. Please try again.")
+				}
+			}
+
+			function add_recommended_node(type = 'routing_node') {
+				const node = recommendations[type];
+				if (!node) {
+					alert("Something went wrong. Please try again.");
```
Comment (DanGould) [inline public/index.html:157]: Can we explain what went wrong and the best recourse for someone seeing the error

File: public/index.html (hunk 4 above)
Comment (0xBEEFCAF3) [reply public/index.html:116]: Wait we can remove the channels var entirely?

File: public/index.html (hunk 6 above)
Comment (0xBEEFCAF3) [reply public/index.html:137]: Lets save this for a follow up PR. https://github.com/chaincase-app/nolooking/issues/66

File: src/http.rs (hunk 9 above)
Comment (0xBEEFCAF3) [reply src/http.rs:133]: Whats the best way to handle this? 
Match the error and return Error(...)?

File: public/style.css (hunk 8 above)
Comment (0xBEEFCAF3) [reply public/style.css:130]: I actually didnt know we had a css framework. We should put a mention in the README or contributing.md.

File: src/recommend.rs (hunk 7 above)
Comment (0xBEEFCAF3) [reply src/recommend.rs:33]: Renames are changing variable name casing.

Comment (0xBEEFCAF3) [reply src/recommend.rs:33]: camel case to snake case

File: src/recommend.rs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    pub edge_node: Node,
+}
+
+// TODO return result here
+async fn body_to_string(body: Body) -> String {
+    let body_bytes = hyper::body::to_bytes(body).await.unwrap();
+    String::from_utf8(body_bytes.to_vec()).unwrap()
+}
+
+pub async fn get_recommended_channels() -> Result<Recommendations, hyper::http::Error> {
```
[the remark this answers is not in the document]
Comment (0xBEEFCAF3) [reply src/recommend.rs:40]: Thanks 👍

File: src/recommend.rs (hunk 7 above)
Comment (0xBEEFCAF3) [reply src/recommend.rs]: fixed in [a897962](https://github.com/chaincase-app/nolooking/pull/51/commits/a8979622e434c1c396f1aa5f60628f1fa8658e53)

File: public/index.html (hunk 5 above)
Comment (0xBEEFCAF3) [reply public/index.html:116]: Thanks, fixed in [b9790f0](https://github.com/chaincase-app/nolooking/pull/51/commits/b9790f0766e757564b6174a2935bdfb0667e3ad7)

File: public/index.html (hunk 10 above)
Comment (nickfarrow) [reply public/index.html:161]: Maybe just something like "Can't connect to node sources right now."
